### ISSUE TITLE
Custom URL routes

### ISSUE THREAD

**Issue by jozefpinter** [blockonomics/woocommerce-plugin#154]
Hello,

my payment URL looks like

> http://localhost/wc-api/WC_Gateway_Blockonomics/?show_order=1C8965rf2BP44Krtm2imutAXAfgFwt9Lth

and I want something like this

> http://localhost/crypto/?show_order=1C8965rf2BP44Krtm2imutAXAfgFwt9Lth

Replace **/wc-api/WC_Gateway_Blockonomics/** with **/crypto/**. That's all.

My **functions.php** looks like

```
add_action( 'init', 'wpse_rewrites_init' );

function wpse_rewrites_init(){
    add_rewrite_rule(
        'crypto/?show_order=([A-Za-z0-9]+)/?$',
      //'wc-api=WC_Gateway_Blockonomics&show_order=$matches[1]',
        'wc-api=WC_Gateway_Blockonomics&show_order=$1',
        'top' );
}
```

Somebody on the internet suggested that **$matches[1]** doesn't work for external redirects, only internal. But anyways, it doesn't work for me.

I was told to consult this issue with Blockonomics dev Support.

**Comment by blockonomics** [comment]
Thanks for taking out the time to try out our plugin and logging this. We are looking at this on priority

**Comment by DarrenWestwood** [comment]
Hi @jozefpinter 

I am having success using the below code:
```
/* Blockonomics Rewrite Rules */
add_action('init', 'blockonomics_rewrite_rules');
function blockonomics_rewrite_rules() {
    add_rewrite_rule( 'crypto/?$', 'index.php?wc-api=WC_Gateway_Blockonomics', 'top' );
}
```
You will now be able to access the checkout page from 
`http://localhost/crypto/?show_order=1C8965rf2BP44Krtm2imutAXAfgFwt9Lth`

However, more changes will still be required to forward the user to this new URL. I will update you here with this information, when available.

**Comment by jozefpinter** [comment]
Hello @DarrenWestwood ,

exactly as you are saying. With your code I can access payment URL via new URL.

But I need get of rid of old URL because payment gateway is sending me there. So **/wc-api/WC_Gateway_Blockonomics/** is still visible.

**Comment by DarrenWestwood** [comment]
In the callback file [here](https://github.com/blockonomics/woocommerce-plugin/blob/master/php/WC_Gateway_Blockonomics.php#L142), we can add the following line:
`$order_url = str_replace('wc-api/WC_Gateway_Blockonomics', 'crypto', $order_url);`
This will now redirect the users to the new URL during the checkout process.

Please note, this change will need to be applied after updating the Blockonomics plugin.

**Comment by jozefpinter** [comment]
@DarrenWestwood ,

can you update the callback in the plugin to redirect users to nicer URL (bitcoin, btc, crypto, whatever) instead  **/wc-api/WC_Gateway_Blockonomics/** ?

Because I think this is really weird URL. It will help to the plugin.

**Comment by DarrenWestwood** [comment]
@jozefpinter, Could you verify the above changes are working correctly for you as well?

We are assessing this option further for possible inclusion into a later release. 
Thank you for this suggestion.

**Comment by jozefpinter** [comment]
@DarrenWestwood , do you mean the code which goes into the file **functions.php** ? It is working fine, but you know... it not helps because plugin is sending me to **/wc-api/WC_Gateway_Blockonomics/** 

or the second snippet of code which goes into the plugin? I didnt try it. Because if you update the plugin I will lose my own changes.

**Comment by DarrenWestwood** [comment]
As we are still accessing the viability and compatibility of the changes for all our users, this integration will require more testing before being applied to plugin releases. 
If you require this feature instantly, you can apply the second snippet to the plugin directly (This will need to be done after updating the plugin).
We would like you to test these new changes to validate they are working as desired.

**Comment by jozefpinter** [comment]
@DarrenWestwood , my website is on localhost. I can't test it.

**Comment by DarrenWestwood** [comment]
You are welcome to download the plugin with the above changes included from here:
https://github.com/blockonomics/woocommerce-plugin/tree/custom-url-routes

or edit the plugin from the Wordpress Plugin Editor:
![Edit Plugins ‹ Wordpress — WordPress](https://user-images.githubusercontent.com/39666372/66437439-041c4b80-ea2a-11e9-95fa-fa81efa46c6e.png)

**Comment by blockonomics** [comment]
>  I didnt try it. Because if you update the plugin I will lose my own changes.

Yes we understand. Just to clarify we want to test if changes @DarrenWestwood  mentioned are working and are seriously considering adding this to the main plugin code soon so that you don't lose this feature on plugin update

**Comment by jozefpinter** [comment]
@DarrenWestwood ,

I've added

> $order_url = str_replace('wc-api/WC_Gateway_Blockonomics', 'crypto', $order_url);

via Plugin editor as you mentioned.

Now it is sending me at new URL. That's everyghing what I can do right now. Because my website is on localhost (offline).

**Comment by DarrenWestwood** [comment]
@jozefpinter 

Both snippets of code need to be applied for this change to work as desired.
1) The `add_rewrite_rule` must be added, which you mention was working correctly to display the page via the new URL.
2) `$order_url = str_replace('wc-api/WC_Gateway_Blockonomics', 'crypto', $order_url);` added to the plugin, to redirect users to the new page.

The plugin [here](https://github.com/blockonomics/woocommerce-plugin/tree/custom-url-routes) has both changes applied in the plugin and is working correctly for me on localhost. 
![Screenshot from 2019-10-10 15-11-27](https://user-images.githubusercontent.com/39666372/66572173-c249ed00-eb70-11e9-8050-3c61755daec3.png)


> I click "Test setup" then "Save". It doesn't help.

Are you receiving an error message when clicking Test Setup?

**Comment by DarrenWestwood** [comment]
Could you also visit Wordpress Dashboard > Settings > Permalink > and Click Save Changes.

This will [flush rewrite rules](https://codex.wordpress.org/Function_Reference/flush_rewrite_rules) for your website, in case other changes you may have made to rewrite rules are causing the issue.

**Comment by jozefpinter** [comment]
@DarrenWestwood ,

1. Thanks, now it works. But I swear I tried it yesterday many times and it not worked. (test setup + save + flush rewrite rules) Today it is working.

2. OK. I will let you know.

**Comment by DarrenWestwood** [comment]
Thank you @jozefpinter for assisting in testing this feature.

**Comment by blockonomics** [comment]
Thanks @jozefpinter we will try to add this in the upcoming release

**Comment by jozefpinter** [comment]
@DarrenWestwood ,

It seems that it is working correctly for me. I've sent some test BTC. I had no issue. But remember, I am on localhost. This is not real environment.

**Comment by blockonomics** [comment]
This merge is deferred until we have a more time / is requested more. Please add a list of test cases to the ticket

**Comment by anktd** [comment]
niche feature request, workaround exists via rewrite rules, closing
